### DISCUSSION TITLE
insert -> Keyerror: 0

### DISCUSSION THREAD

**Discussion by CrypTraider** [milvus-io/milvus#23916] [Q&A and General discussion] [answered]
I am STUMPED. I have checked and double checked the data and I'm still getting this Python error:

Traceback (most recent call last):
  File "D:\Oathlords\AI\Python\new_chat.py", line 101, in <module>
    similar_chats = process_chat(tick, text, recipient, sender, omit_record_ids, memory_count)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\Oathlords\AI\Python\new_chat.py", line 71, in process_chat
    collection.insert([new_chat])
  File "C:\Python311\Lib\site-packages\pymilvus\orm\collection.py", line 424, in insert
    check_insert_data_schema(self._schema, data)
  File "C:\Python311\Lib\site-packages\pymilvus\orm\schema.py", line 312, in check_insert_data_schema
    infer_fields = parse_fields_from_data(data)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\site-packages\pymilvus\orm\schema.py", line 343, in parse_fields_from_data
    fields = [FieldSchema("", infer_dtype_bydata(d[0])) for d in data]
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\site-packages\pymilvus\orm\schema.py", line 343, in <listcomp>
    fields = [FieldSchema("", infer_dtype_bydata(d[0])) for d in data]
                                                 ~^^^
KeyError: 0

This is the simplified code using pymilvus 2.2.8. The error happens on the last line:

```
def process_chat(tick, text, recipient, sender):
    field_types = [
        FieldSchema(name="tick", dtype=DataType.INT64, is_primary=True, auto_id=False),
        FieldSchema(name="sender", dtype=DataType.VARCHAR, max_length=32),
        FieldSchema(name="recipient", dtype=DataType.VARCHAR, max_length=32),
        FieldSchema(name="vector", dtype=DataType.FLOAT_VECTOR, dim=1024),
        FieldSchema(name="text", dtype=DataType.VARCHAR, max_length=256)
    ]
    schema = CollectionSchema(fields=field_types, description="All chats.")
    collection = Collection(name="Chats", schema=schema)

    index_params = {
        "metric_type": "L2",
        "index_type": "IVF_FLAT",
        "params": {"nlist": 1024}
    }
    collection.create_index("vector", index_params)
    collection.load()

    # Get the embeddings from BERT.
    input_ids = tokenizer.encode(text, return_tensors="pt")
    with torch.no_grad():
        output = model(input_ids, output_hidden_states=True)
        last_hidden_state = output.hidden_states[-1]
        embedding = last_hidden_state[0].mean(dim=0).numpy()

    new_chat = {
        "tick": tick,
        "sender": sender,
        "recipient": recipient,
        "vector": embedding.tolist(),
        "text": text
    }

    collection.insert([new_chat])
```

This is the test data. There are 1024 elements in vector:

{'tick': 638189834034804559, 'sender': 'Thrull_Test', 'recipient': 'Cutha_Test', 'vector': [-0.23536615073680878, ..., -0.05048232898116112, -0.23860548436641693, -0.03905796632170677], 'text': 'dsasdfasdfasdfasdf'}

**Comment by CrypTraider** [comment]
This AI stuff is moving so fast that I start to panic when I lose a whole day of work, lol. I still can't find this problem. I can't imagine how this could be an issue with Milvus since it's so simple and no one else is having the issue.... but I can't find any other problem. It implies the first element doesn't exist, but it clearly exists.

**Comment by CrypTraider** [comment]
*throws his monitor out the window*
*pulls hair out*
*shoots self in face*

**Comment by yhmo** [comment, marked as answer]
The insert() api requires a "list of list".
Your collection has 5 fields, the correct way is:

```
    new_chat = [
        [638189834034804559],                     # tick field, a list of int64
        ['Thrull_Test'],                          # sender field, a list of varchar
        ['Cutha_Test'],                           # recipient field, a list of varchar
        [ [-0.23536615073680878, ..., -0.05048232898116112, -0.23860548436641693, -0.03905796632170677] ]   # vector field, it is a list of list
        ['dsasdfasdfasdfasdf'],                   # text field, a list of varchar
    ]

    collection.insert(new_chat)
```

Example:
https://github.com/milvus-io/pymilvus/blob/6bb23e37715392f9398e5b289cc0a4f2667fc563/examples/example_str.py#L79

**Comment by CrypTraider** [comment]
I knew that ;)

You're the best person in the entire world today. You should celebrate by conquering Manchuria. I hear it's got a nice coastline.

Thanks!
